### PR TITLE
APR-1477 RoATP FHA functional & design update

### PR DESCRIPTION
**Pull request by martin-hornsby**

### CODE REVIEW

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs [hunk 1]
```diff
@@ -200,28 +194,92 @@ private async Task<RoatpFinancialApplicationViewModel> CreateRoatpFinancialAppli
...
+            var financialSections = await GetFinancialSections(applicationFromAssessor.ApplicationId);
+
+            return new RoatpFinancialApplicationViewModel(applicationFromAssessor, parentCompanySection, activelyTradingSection, organisationTypeSection, financialSections);
+        }
+
+        private async Task<Section> GetParentCompanySection(Guid applicationId)
+        {
+            const string ParentCompanySectionTitle = "UK ultimate parent company";
+            const string HasParentCompanyQuestionTag = "HasParentCompany";
+            const string HasParentCompanyPageId = "20";
```
Comment (David-B-Read) [inline src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: Just a heads up, in the APR-1275 branch there's a constants file for all the page ids. It's being reused by assessor so makes sense to use it here too

https://github.com/SkillsFundingAgency/das-admin-service/blob/0b8ec5558fe7f5a2aaac895c92e7cc9882e90e94/src/SFA.DAS.AssessorService.ApplyTypes/Roatp/RoatpQnaConstants.cs

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs [hunk 2]
```diff
@@ -200,28 +205,107 @@ private async Task<RoatpFinancialApplicationViewModel> CreateRoatpFinancialAppli
...
                     var roatpSequence = roatpSequences.FirstOrDefault(x => x.Id == sequence.Id);
 
                     if (!roatpSequence.ExcludeSections.Contains(section.SectionNo.ToString()))
                     {
-                        financialSections.Add(section);
+                        // Ensure at least one active page is answered
+                        if (section.QnAData.Pages.Any(p => p.Active && p.Complete))
+                        {
+                            // APR-1477 - adjust title for Assessor
+                            if(section.SequenceNo == FinancialHealthSequenceNo && section.SectionNo == 2)
```
Comment (David-B-Read) [inline src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: Can we replace magic section numbers with constants, and have them all in a logical place?

File: src/SFA.DAS.AdminService.Web/ViewModels/Roatp/Financial/RoatpFinancialApplicationViewModel.cs [hunk 3]
```diff
@@ -71,7 +100,46 @@ private void SetupGradeAndFinancialDueDate(FinancialReviewDetails financialRevie
...
+            {
+                DeclaredInApplication = fhaSequence.NotRequired ? "Exempt" : "Not exempt";
+            }
+        }
+
+        public string GetDownloadFilesLinkText(int sequenceNo, int sectionNo)
+        {
+            string linkText;
+
+            if (sequenceNo == 2 && sectionNo == 2)
```
Comment (David-B-Read) [inline src/SFA.DAS.AdminService.Web/ViewModels/Roatp/Financial/RoatpFinancialApplicationViewModel.cs]: Replace with constants please

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs (hunk 1 above)
Comment (MarkFCain) [reply src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: the gateway-submaster might have an even longer list, so if you lift it, look in there and compare

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs [hunk 4]
```diff
@@ -200,28 +205,107 @@ private async Task<RoatpFinancialApplicationViewModel> CreateRoatpFinancialAppli
...
+            return activelyTradingSection;
+        }
+
+        private async Task<Section> GetOrganisationTypeSection(Guid applicationId)
+        {
+            const string OrganisationTypeSectionTitle = "Organisation type";
+            const string OrganisationTypeMainSupportingPageId = "140";
+            const string OrganisationTypeEmployerPageId = "150";
+
+            Section organisationTypeSection = await _qnaApiClient.GetSectionBySectionNo(applicationId, OrganisationDetailsSequenceNo, 4);
```
Comment (MarkFCain) [inline src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: magic number

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs [hunk 5]
```diff
@@ -200,28 +205,107 @@ private async Task<RoatpFinancialApplicationViewModel> CreateRoatpFinancialAppli
...
+            const string HasParentCompanyPageId = "20";
+            const string ParentCompanyDetailsPageId = "21";
+
+            Section parentCompanySection = null;
+
+            var hasParentCompanyTagValue = await _qnaApiClient.GetQuestionTag(applicationId, HasParentCompanyQuestionTag);
+
+            if ("Yes".Equals(hasParentCompanyTagValue, StringComparison.OrdinalIgnoreCase))
+            {                
+                parentCompanySection = await _qnaApiClient.GetSectionBySectionNo(applicationId, OrganisationDetailsSequenceNo, 2);
```
Comment (MarkFCain) [inline src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: replace magic number please

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs [hunk 6]
```diff
@@ -200,28 +205,107 @@ private async Task<RoatpFinancialApplicationViewModel> CreateRoatpFinancialAppli
...
+        }
+
+        private async Task<Section> GetActivelyTradingSection(Guid applicationId)
+        {
+            const string ActivelyTradingSectionTitle = "Actively trading";
+            const string TradingForMainPageId = "50";
+            const string TradingForEmployerPageId = "51";
+            const string TradingForSupportingPageId = "60";
+
+            Section activelyTradingSection = await _qnaApiClient.GetSectionBySectionNo(applicationId, OrganisationDetailsSequenceNo, 2);
```
Comment (MarkFCain) [inline src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: replace magic number please

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs [hunk 7]
```diff
@@ -272,6 +356,24 @@ private async Task<List<FinancialEvidence>> GetFinancialEvidence(Guid applicatio
...
+            switch (vm?.FinancialReviewDetails?.SelectedGrade)
+            {
+                case FinancialApplicationSelectedGrade.Clarification:
+                    return vm.ClarificationComments;
+                case FinancialApplicationSelectedGrade.Inadequate:
+                    return vm.InadequateComments;
+                default:
+                    return null;
+            }
+        }
```
Comment (MarkFCain) [inline src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: Needs tests adding

File: src/SFA.DAS.AdminService.Web/Validators/Roatp/Applications/RoatpFinancialApplicationViewModelValidator.cs [hunk 8]
```diff
@@ -80,5 +83,21 @@ private void ProcessDate(FinancialDueDate dueDate, string propertyName, CustomCo
                 context.AddFailure(propertyName, "Financial due date must be a future date");
             }
         }
+
```
Comment (MarkFCain) [inline src/SFA.DAS.AdminService.Web/Validators/Roatp/Applications/RoatpFinancialApplicationViewModelValidator.cs]: Validator needs coverage

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs (hunk 1 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: copied across from gateway-submaster :)

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs (hunk 2 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: done

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs (hunk 4 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: sorted

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs (hunk 5 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: sorted

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs (hunk 6 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpFinancialController.cs]: sorted

File: src/SFA.DAS.AdminService.Web/ViewModels/Roatp/Financial/RoatpFinancialApplicationViewModel.cs (hunk 3 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AdminService.Web/ViewModels/Roatp/Financial/RoatpFinancialApplicationViewModel.cs]: done

File: src/SFA.DAS.AdminService.Web/ViewModels/Roatp/Financial/RoatpFinancialApplicationViewModel.cs [hunk 9]
```diff
@@ -71,7 +100,46 @@ private void SetupGradeAndFinancialDueDate(FinancialReviewDetails financialRevie
...
+            else if (FinancialReviewDetails.SelectedGrade == FinancialApplicationSelectedGrade.Clarification)
+            {
+                ClarificationComments = FinancialReviewDetails.Comments;
+            }
         }
-    }
 
+        private void SetupDeclaredInApplication(RoatpApplyData applyData)
+        {
+            const int fhaSequenceNo = 2;
```
Comment (David-B-Read) [inline src/SFA.DAS.AdminService.Web/ViewModels/Roatp/Financial/RoatpFinancialApplicationViewModel.cs]: Can RoatpQnaConstants.RoatpSequences.FinancialEvidence be used here?

Comment (martin-hornsby) [reply src/SFA.DAS.AdminService.Web/ViewModels/Roatp/Financial/RoatpFinancialApplicationViewModel.cs]: done

File: src/SFA.DAS.AdminService.Web/Validators/Roatp/Applications/RoatpFinancialApplicationViewModelValidator.cs (hunk 8 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AdminService.Web/Validators/Roatp/Applications/RoatpFinancialApplicationViewModelValidator.cs]: done - added an exhaustive set of tests

**Pull request merged**